feat: Support ContactRect in X11 and Windows platform

lindexi opened the pull request:
<!--- See CONTRIBUTING.md for general guidelines on contributions -->

## What does the pull request do?
<!--- Give a bit of background on the PR here, together with links to with related issues etc. -->
Adding the touch size support, which know as [ContactRect](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.input.pointerpointproperties.contactrect?view=windows-app-sdk-1.5 ) in X11 and Windows platform.

## What is the current behavior?
<!--- If the PR is a fix, describe the current incorrect behavior, otherwise delete this section. -->
Miss the touch size feature.

## What is the updated/expected behavior with this PR?
<!--- Describe how to test the PR. -->

We can get the touch size from X11 and Windows.

I test the Avalonia Windows version by my two demo, the [TouchSizeAvalonia demo](https://github.com/lindexi/lindexi_gd/commit/4c5a4b06b3c3aba549c2e3cbc4f319185a8d337e) and the [WinUI demo](https://github.com/lindexi/lindexi_gd/tree/03dbd1234158418bbff757461c24e51029e068ca/AvaloniaIDemo/TouchSizeAvalonia/LilajibereKawerfelnarrarcu). 

I use the same physics for touch, and I can get the same result from  [TouchSizeAvalonia demo](https://github.com/lindexi/lindexi_gd/commit/4c5a4b06b3c3aba549c2e3cbc4f319185a8d337e) and the [WinUI demo](https://github.com/lindexi/lindexi_gd/tree/03dbd1234158418bbff757461c24e51029e068ca/AvaloniaIDemo/TouchSizeAvalonia/LilajibereKawerfelnarrarcu).  That means this PR can have the same implementation as WinUI. Here's a video I recorded on my phone:

[TouchSize.mp4](https://github.com/user-attachments/assets/bf6db648-f84f-4889-98d0-fc49bc16cfaa)


## How was the solution implemented (if it's not obvious)?
<!--- Include any information that might be of use to a reviewer here. -->


## Checklist

- [ ] Added unit tests (if possible)? I test pass with my [demo](https://github.com/lindexi/lindexi_gd/commit/4c5a4b06b3c3aba549c2e3cbc4f319185a8d337e)
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes
<!--- List any breaking changes here. -->
Add the ContactRect property to PointerPointProperties and changed the argument list in the constructor of PointerPointProperties.



## Obsoletions / Deprecations
<!--- Obsolete and Deprecated attributes on APIs MUST only be included when discussed with Core team. @grokys, @kekekeks & @danwalmsley -->

## Fixed issues
<!--- If the pull request fixes issue(s) list them like this: 
Fixes #123
Fixes #456
-->
Reference: https://github.com/AvaloniaUI/Avalonia/issues/1477

lindexi committed:
feat: Support ContactRect in X11 platform

kekekeks commented on src/Avalonia.Base/Input/PointerPoint.cs:
```diff
@@ -155,17 +160,18 @@ public PointerPointProperties(RawInputModifiers modifiers, PointerUpdateKind kin
 
         public PointerPointProperties(RawInputModifiers modifiers, PointerUpdateKind kind,
-            float twist, float pressure, float xTilt, float yTilt
+            float twist, float pressure, float xTilt, float yTilt, Rect contactRect
```
Why do we need to extend the API? Can't relevant XI2 properties be converted to existing ones? How other platforms expose the same information? How browsers are handling that?

kekekeks replied:
How the same info is exposed by UWP?

lindexi replied:
@kekekeks It can be implemented in Windows platform, see https://learn.microsoft.com/en-us/windows-hardware/design/component-guidelines/touchscreen-required-hid-top-level-collections

lindexi replied:
Also, we can get the touch size in wpf by [`StylusPointProperties.Width`](https://learn.microsoft.com/en-us/dotnet/api/system.windows.input.styluspointproperties.width) and [`StylusPointProperties.Height`](https://learn.microsoft.com/en-us/dotnet/api/system.windows.input.styluspointproperties.height)

lindexi replied:
> How other platforms expose the same information? How browsers are handling that?

I support the windows platform now.

kekekeks commented on src/Avalonia.Base/Input/PointerPoint.cs:
```diff
@@ -155,17 +160,18 @@ public PointerPointProperties(RawInputModifiers modifiers, PointerUpdateKind kin
 
         internal PointerPointProperties(RawInputModifiers modifiers, PointerUpdateKind kind, RawPointerPoint rawPoint)
-            : this(modifiers, kind, rawPoint.Twist, rawPoint.Pressure, rawPoint.XTilt, rawPoint.YTilt)
+            : this(modifiers, kind, rawPoint.Twist, rawPoint.Pressure, rawPoint.XTilt, rawPoint.YTilt, rawPoint.ContactRect)
```
We need to provide some compat value (like 1x1 rect) for cases where contact rect isn't supplied by the platform backend.

kekekeks commented:
Also, according to https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.input.pointerpointproperties.contactrect?view=windows-app-sdk-1.5 the contact rect is supposed to be reported in DIPs

kekekeks commented on src/Avalonia.X11/XI2Manager.cs:
```diff
@@ -256,6 +268,52 @@ private void OnDeviceEvent(IXI2Client client, ParsedDeviceEvent ev)
+                        Debug.Assert(client is IWindowImpl);
+                        if (client is IWindowImpl windowImpl)
+                        {
+                            var screen = _platform.X11Screens.ScreenFromWindow(windowImpl);
```
The window can occupy multiple screens, using ScreenFromWindow for determining the position and dimensions of the touch origin screen isn't sufficient.

lindexi replied:
Thank you @kekekeks . I enlisted the help of my [partner](https://github.com/kkwpsv), borrowed another touchscreen, and assembled a device with two touchscreens. I conducted tests on the Kylin (debian bullseye/sid) system. Through my testing, I found that X11 will automatically scale the input touch TouchMajor across multiple screens based on the screen resolution. Therefore, the Monitor obtained is the Monitor where the touch is located, and the resolution of the Monitor can be used to calculate the value of touch pixels. However, the behavior in X11 is very similar to the situation on Windows without PM DPI awareness enabled. Due to the limitations of X11, my code cannot accurately reflect the physical area on multiple touch screens, but this is a limitation of X11. I suggest waiting until other developers have multi-touch screen requirements in the future before discussing this behavior in detail. At least my current code status can work normally on a single touchscreen.

lindexi committed:
Provide compat value for ContactRect property

lindexi committed:
Using touch position to get the screen.

lindexi committed:
Try get the correct screen

lindexi committed:
Remove the code of log.

lindexi committed:
Improve performance avoid the GC stress.

maxkatz6 commented on src/Avalonia.Base/Input/Raw/RawPointerEventArgs.cs:
```diff
@@ -143,6 +143,14 @@ public record struct RawPointerPoint
+        /// <inheritdoc cref="PointerPointProperties.ContactRect" />
+        public Rect ContactRect
+        {
+            get => _contactRect ?? new Rect(Position, new Size());
```
it should be new Size(1,1) by default I believe

lindexi replied:
@maxkatz6 Sorry, I do not think so. I think the size should be zero here as the WinUI do.

maxkatz6 commented on src/Avalonia.Base/Input/PointerPoint.cs:
```diff
@@ -155,17 +160,18 @@ public PointerPointProperties(RawInputModifiers modifiers, PointerUpdateKind kin
 
         public PointerPointProperties(RawInputModifiers modifiers, PointerUpdateKind kind,
-            float twist, float pressure, float xTilt, float yTilt
+            float twist, float pressure, float xTilt, float yTilt, Rect contactRect
```
Adding a new parameter breaks public API. You can either add a second overload, or use `public Rect ContactRect { get; init; }`

lindexi replied:
Address.

lindexi committed:
Add a second overload in PointerPointProperties to avoid break API change

lindexi committed:
Merge remote-tracking branch 'origin/master' into t/lindexi/TouchMajor

lindexi commented:
Thanks https://github.com/AvaloniaUI/Avalonia/pull/16705

I can call the `_platform.Screens.ScreenFromPoint`.

lindexi committed:
Support touch size in WM_Touch

lindexi committed:
Support get ContactRect in WM_POINTER

lindexi committed:
Fix the coordinate of touch contact area

lindexi commented on src/Windows/Avalonia.Win32/WindowImpl.AppWndProc.cs:
```diff
@@ -436,14 +436,34 @@ protected virtual unsafe IntPtr AppWndProc(IntPtr hWnd, uint msg, IntPtr wParam,
+                                            (touchInput.Y + touchInput.CyContact) / 100);
+                                    var bottomRightPosition = PointToClient(bottomRightPixelPoint);
+
+                                    rawPointerPoint.ContactRect = new Rect(position, bottomRightPosition);
```
The position is the center of the contact area. This is wrong.

lindexi committed:
Fix the calcuate of WM_Touch contact area

lindexi committed:
Fix the X11 area contact

lindexi committed:
Fix the touch major and minor default value is not null.

The XIValuatorClassInfo is struct, so the FirstOrDefault will return the default struct when not found.

maxkatz6 commented on src/Windows/Avalonia.Win32/WindowImpl.AppWndProc.cs:
```diff
@@ -436,14 +436,46 @@ protected virtual unsafe IntPtr AppWndProc(IntPtr hWnd, uint msg, IntPtr wParam,
                         {
                             foreach (var touchInput in touchInputs)
                             {
+                                var position = PointToClient(new PixelPoint(touchInput.X / 100, touchInput.Y / 100));
```
Feel free to drop WM_TOUCH handling for ContactRect, as it's not going to be used anywhere on modern devices. WM_POINTER is the main way we handle touch/pen input.

lindexi replied:
@maxkatz6 There is the other issues in avalonia, that the RegisterTouchWindow will be call event enable the WM_POINTER.

And the win7 system still use the WM_TOUCH now.

maxkatz6 replied:
We don't support win7 at this point.

maxkatz6 replied:
To clarify, we try to keep Avalonia compatible with Win7, but we don't go extra miles to support some specific features. And if some code is only used on Win7, we don't really need it, as it only adds maintenance cost.

maxkatz6 reviewed (approved):
Aside from unnecessary `WM_TOUCH` handling, Windows implementation and API looks good to me.

maxkatz6 commented:
Merging for 11.3 master branch. I will try to push more implementations over next months before 11.3 release.

lindexi commented:
Thank you @maxkatz6

lindexi commented on src/Windows/Avalonia.Win32/WindowImpl.AppWndProc.cs:
```diff
@@ -1053,13 +1085,35 @@ private RawPointerPoint CreateRawPointerPoint(POINTER_TOUCH_INFO info)
+
+                // Why not use ptPixelLocationX and ptPixelLocationY to as leftTopPosition?
+                // Because ptPixelLocationX and ptPixelLocationY will be the center of the contact area.
+                pointerPoint.ContactRect = new Rect(leftTopPosition, bottomRightPosition);
```
Addition: The behavior here is identical to that of WinUI 3, even when the DPI is not set to 100%.
